Code/Framework/AzCore/AzCore/IO/Streamer/Statistics.h

```diff
@@ -10,127 +10,297 @@
 
-            //! Calculates the average value for the window.
-            AverageType CalculateAverage() const
+    private:
+        Statistic() = default;
+
+        template<typename T>
+        static Statistic Create(
+            AZStd::string_view owner, AZStd::string_view name, T&& value, AZStd::string_view description, GraphType graphType);
+
+        Value m_value;
+        AZStd::string_view m_owner;
+        AZStd::string_view m_name;
+        AZStd::string_view m_description;
+        GraphType m_graphType;
+    };
+
+    //! AverageWindow keeps track of the average of values in a sliding window.
+    //! @StorageType The type of the value in the sliding window. Need to be a number of a AZStd::chrono::duration.
+    //! @AverageType The type CalculateAverage will return. StorageType needs to be able to be converted to AverageType.
+    //! @WindowSize The maximum number of entries kept in the window.
+    template<typename StorageType, typename AverageType, size_t WindowSize>
+    class AverageWindow
+    {
+        static_assert(AZStd::is_arithmetic<StorageType>::value || AZStd::chrono::Internal::is_duration<StorageType>::value,
+            "AverageWindow only support numbers and AZStd::chrono::durations."); 
+        static_assert((WindowSize & (WindowSize - 1)) == 0, "The WindowSize of AverageWindow needs to be a power of 2.");
+    public:
+        static const size_t s_windowSize = WindowSize;
+
+        AverageWindow()
+        {
+            memset(&m_values, 0, sizeof(m_values));
+            memset(&m_runningTotal, 0, sizeof(m_runningTotal));
+            if constexpr (AZStd::is_same_v<StorageType, Statistic::TimeValue>)
             {
-                size_t count = m_count > 0 ? m_count : 1;
-                return static_cast<AverageType>(static_cast<AverageType>(m_runningTotal) / static_cast<AverageType>(count < WindowSize ? count : WindowSize)); 
+                m_minValue = Statistic::TimeValue::max();
+                m_maxValue = Statistic::TimeValue::min();
             }
-            //! Gets the running total of the values in the window.
-            StorageType GetTotal() const { return m_runningTotal; }
-            //! Returns the total number of entries that have been passed in. This is not the total number of entries that are stored however.
-            size_t GetNumRecorded() const { return m_count; }
-
-        private:
-            StorageType m_values[WindowSize];
-            StorageType m_runningTotal;
-            size_t m_count = 0;
-        };
+        }
 
-        using TimedAverageWindowDuration = AZStd::chrono::microseconds;
-        template<size_t WindowSize>
-        using TimedAverageWindow = AverageWindow<TimedAverageWindowDuration, TimedAverageWindowDuration, WindowSize>;
+        //! Push a new entry into the window. If the window is full, the oldest value will be removed.
+        void PushEntry(StorageType value)
+        {
+            size_t index = m_count & (WindowSize - 1);
```

Comment:
Any reason to use the template parameter `WindowSize` here instead of the constant `s_windowSize`?

Reply:
Not really. They both mean the same thing and have the same functionality. Storing the `WindowSize` in `s_windowSize` is mostly there in case another piece of code interested in knowing the window size.